### PR TITLE
Use headers for metadata

### LINKED ISSUE #14: Remove use of `method_missing`
I added this as a convenience, but it is really a code smell. This should be removed: https://github.com/simplybusiness/rabbit_feed/blob/master/lib/rabbit_feed/event.rb#L33-L35

### PR DESCRIPTION
**Pull request by joshuafleck**
@calo81 This PR fixes this issue that you raised: https://github.com/simplybusiness/rabbit_feed/issues/10, it also fixes #14 

The practice of pushing the event metadata into the payload was ugly - It is much cleaner now that the event metadata is passed in the message headers.

@calo81 @dawid-sklodowski @martinbtt @tpaul09 Could you please sign off?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -166,20 +166,20 @@ describe 'consuming events' do
...
   end
 
-  let(:event) { {'application' => 'app', 'name' => 'ev', 'stuff' => 'some_stuff'} }
+  let(:payload) { {'stuff' => 'some_stuff'} }
 
   before { define_route }
 
   it 'route to the correct service' do
-    rabbit_feed_consumer.consume_event(event)
+    rabbit_feed_consumer.consume_event('event_name', 'application_name', payload)
```
Comment (calo81) [inline README.md]: I think I would prefer this to still be `consume_event(event)`.. cleaner interface and event has a `event.headers` and `event.payload`.. then the `consume_event` would extract those

File: example/non_rails_app/spec/lib/non_rails_app/event_routing_spec.rb [hunk 2]
```diff
@@ -2,11 +2,11 @@
...
 module NonRailsApp
   describe 'Event Routing' do
-    let(:event) { {'application' => 'rails_app', 'name' => 'user_creates_beaver'} }
+    let(:payload) { { 'field' => 'value' } }
 
     it 'routes events correctly' do
-      expect(NonRailsApp::EventHandler).to receive(:handle_event) { |full_event| expect(full_event.payload).to eq(event)}
-      rabbit_feed_consumer.consume_event(event)
+      expect(NonRailsApp::EventHandler).to receive(:handle_event) { |full_event| expect(full_event.payload).to eq(payload)}
+      rabbit_feed_consumer.consume_event('user_creates_beaver', 'rails_app', payload)
```
Comment (calo81) [inline example/non_rails_app/spec/lib/non_rails_app/event_routing_spec.rb]: same here as https://github.com/simplybusiness/rabbit_feed/pull/16/files#r24749194

File: example/rails_app/spec/event_routing_spec.rb [hunk 3]
```diff
@@ -2,11 +2,11 @@
 
 module RailsApp
   describe 'Event Routing' do
-    let(:event) { {'application' => 'non_rails_app', 'name' => 'application_acknowledges_event'} }
+    let(:payload) { {'field' => 'value'} }
 
     it 'routes events correctly' do
       expect(::EventHandler).to receive(:handle_event)
-      rabbit_feed_consumer.consume_event(event)
+      rabbit_feed_consumer.consume_event('application_acknowledges_event', 'non_rails_app', payload)
```
Comment (calo81) [inline example/rails_app/spec/event_routing_spec.rb]: https://github.com/simplybusiness/rabbit_feed/pull/16/files#r24749194

File: lib/rabbit_feed/consumer.rb [hunk 4]
```diff
@@ -5,8 +5,8 @@ module Consumer
     attr_accessor :event_routing
 
     def run
-      ConsumerConnection.consume do |raw_event|
-        event = Event.deserialize raw_event
+      ConsumerConnection.consume do |payload, metadata|
```
Comment (calo81) [inline lib/rabbit_feed/consumer.rb]: Again, I would prefer that the `event` or `raw_event` encapsulate the payload an metadata (which I probably would call headers as is a more common term in messaging)... as they are not even needed separate here and `Event.deserialize raw_event` looks better

File: lib/rabbit_feed/event.rb [hunk 5]
```diff
@@ -2,12 +2,19 @@ module RabbitFeed
...
+    validates :schema, :application, :host, :environment, :created_at_utc, :version, :name, presence: true
+    validates :payload, length: { minimum: 0, allow_nil: false, message: 'can\'t be nil' }
 
-    def initialize schema, payload
-      @schema  = schema
-      @payload = payload
+    def initialize schema, payload, metadata
+      @schema         = schema
+      @payload        = payload.with_indifferent_access.freeze if payload
+      @application    = metadata['application']
```
Comment (calo81) [inline lib/rabbit_feed/event.rb]: the event for me would have a single `headers` property that could point to anoither `Header` object that has these properties on it. then you can get them with `event.headers`

File: lib/rabbit_feed/event.rb [hunk 6]
```diff
@@ -19,25 +26,21 @@ def serialize
       buffer.string
     end
 
-    def self.deserialize event
+    def self.deserialize serialized_payload, metadata
```
Comment (calo81) [inline lib/rabbit_feed/event.rb]: where is `metadata` used in this method?

File: lib/rabbit_feed/publishing_options.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+module RabbitFeed
+  class PublishingOptions
```
Comment (calo81) [inline lib/rabbit_feed/publishing_options.rb]: I am not sure I understand this object very well. I think it holds the headers for a while... so maybe this is nonsense but Why it needs to live on its own as it seems to be 1 to 1 related to the event and make no sense without it... Can't it be created when creating the event?

`Event.new name, payload, event_definition` and then extract this object (if needed) internally?

**Comment by calo81** [comment]
I think most of the things I wrote just pretend to mean that the `headers` for me are part of the message, just not part of the content payload. So I would expect and instance of `Event` to give me its headers. and a `serialized_event` to contain them as well. Something like: http://www.mulesoft.org/documentation/display/current/Mule+Message+Structure

**Comment by calo81** [comment]
Other thing is that `metadata` and `header` seem to be used interchangeably.. Better to stick with 1... and better `header` :smile:

**Comment by joshuafleck** [comment]
@calo81 - Valid points. My thinking is as follows:

* The term `header` to me implies the means for transferring the information. The term `metadata` describes what this information is. I used the `headers` field provided by rabbit mq as that seemed what the fields was designed for.

I see where you are coming from - One concern I have now is with the metadata in the header, it makes the metadata impossible to version (especially needed for backwards compatibility). I'm going to rethink this. Thanks for the feedback!

**Comment by joshuafleck** [comment]
Closing while I re-think this.

**Comment by joshuafleck** [comment]
@calo81 - Please see updated PR here: https://github.com/simplybusiness/rabbit_feed/pull/17

**Pull request closed without merging**